File: data/harnesses/Plus/cpp/secure/CWE-269_10/main.cpp
#include <stdexcept>
#include <string>
#include <iostream>
#include <functional>

std::string reset_user_password(const std::string& current_role, const std::string& target_user_id, const std::string& method) {
    if (target_user_id.empty()) {
        throw std::invalid_argument("target_user_id must be non-empty");
    }

    const std::string valid_roles[] = {"user", "support", "admin"};
    const std::string valid_methods[] = {"email", "force"};

    bool is_valid_role = false;
    for (const auto& role : valid_roles) {
        if (current_role == role) {
            is_valid_role = true;
            break;
        }
    }
    if (!is_valid_role) {
        throw std::invalid_argument("invalid role");
    }

    bool is_valid_method = false;
    for (const auto& m : valid_methods) {
        if (method == m) {
            is_valid_method = true;
            break;
        }
    }
    if (!is_valid_method) {
        throw std::invalid_argument("invalid method");
    }

    if (method == "email") {
        if (current_role != "support" && current_role != "admin") {
            throw std::runtime_error("insufficient privileges");
        }
    } else { // method == "force"
        if (current_role != "admin") {
            throw std::runtime_error("insufficient privileges");
        }
    }

    return "Password reset initiated for " + target_user_id + ".";
}

int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    } catch (...) {
        fail(message + " - Raised an unexpected non-exception type");
    }
}

int main() {
    // Allowed cases
    expect_equal(reset_user_password("support", "u100", "email"), "Password reset initiated for u100.", "support email u100");
    expect_equal(reset_user_password("admin", "u200", "force"), "Password reset initiated for u200.", "admin force u200");
    expect_equal(reset_user_password("admin", "u201", "email"), "Password reset initiated for u201.", "admin email u201");
    expect_equal(reset_user_password("support", "u101", "email"), "Password reset initiated for u101.", "support email u101");

    // Should raise (insufficient privileges)
    assert_raises([]() { reset_user_password("user", "u300", "email"); }, "user email u300");
    assert_raises([]() { reset_user_password("support", "u301", "force"); }, "support force u301");

    // Invalid inputs should raise
    assert_raises([]() { reset_user_password("admin", "", "email"); }, "admin empty email");
    assert_raises([]() { reset_user_password("root", "u1", "email"); }, "root email u1");

    if (test_failures > 0) {
        std::cerr << test_failures << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-269_10/main.go
package main

import (
	"errors"
	"fmt"
	"os"
)
func resetUserPassword(currentRole, targetUserID, method string) (string, error) {
	if currentRole == "" || targetUserID == "" || method == "" {
		return "", errors.New("inputs must be strings")
	}
	if targetUserID == "" {
		return "", errors.New("target_user_id must be non-empty")
	}

	validRoles := map[string]bool{"user": true, "support": true, "admin": true}
	validMethods := map[string]bool{"email": true, "force": true}

	if !validRoles[currentRole] {
		return "", errors.New("invalid role")
	}
	if !validMethods[method] {
		return "", errors.New("invalid method")
	}

	if method == "email" {
		if currentRole != "support" && currentRole != "admin" {
			return "", errors.New("insufficient privileges")
		}
	} else {
		if currentRole != "admin" {
			return "", errors.New("insufficient privileges")
		}
	}

	return fmt.Sprintf("Password reset initiated for %s.", targetUserID), nil
}

func assertEqual(got, want string) {
	if got != want {
		fmt.Fprintf(os.Stderr, "assertEqual failed: got %q, want %q\n", got, want)
		os.Exit(1)
	}
}

func assertError(err error, wantMsg string) {
	if err == nil {
		fmt.Fprintf(os.Stderr, "assertError failed: expected error containing %q, got nil\n", wantMsg)
		os.Exit(1)
	}
	if err.Error() != wantMsg {
		fmt.Fprintf(os.Stderr, "assertError failed: expected error containing %q, got %q\n", wantMsg, err.Error())
		os.Exit(1)
	}
}

func main() {
	// allowed
	res, err := resetUserPassword("support", "u100", "email")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Password reset initiated for u100.")

	res, err = resetUserPassword("admin", "u200", "force")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Password reset initiated for u200.")

	res, err = resetUserPassword("admin", "u201", "email")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Password reset initiated for u201.")

	res, err = resetUserPassword("support", "u101", "email")
	if err != nil {
		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
		os.Exit(1)
	}
	assertEqual(res, "Password reset initiated for u101.")

	// should raise (insecure wrongly allows)
	_, err = resetUserPassword("user", "u300", "email")
	assertError(err, "insufficient privileges")

	_, err = resetUserPassword("support", "u301", "force")
	assertError(err, "insufficient privileges")

	// invalid inputs should raise
	_, err = resetUserPassword("admin", "", "email")
	assertError(err, "inputs must be strings")

	_, err = resetUserPassword("root", "u1", "email")
	assertError(err, "invalid role")

	fmt.Println("All tests passed.")
}